File: src/main/java/org/frc1778/freezylib/logging/PolledField.java
package org.frc1778.freezylib.logging;

import java.util.function.Supplier;
import org.frc1778.freezylib.util.Measurement;

public class PolledField extends Field {

  private final Supplier<String> supplier;
  private transient String value;

  public <T> PolledField(String name, String unit, Supplier<T> supplier) {
    setName(name);
    setUnit(unit);
    setType(supplier.get().getClass());
    this.supplier = () -> String.valueOf(supplier.get());
  }

  public <T> PolledField(String name, Measurement.Base unit, Supplier<T> supplier) {
    this(name, unit.getSymbol(), supplier);
  }

  public void pollSupplier() {
    value = supplier.get();
  }

  public Supplier<String> getSupplier() {
    return supplier;
  }

  @Override
  public String getValue() {
    return value;
  }
}

File: src/main/kotlin/org/frc1778/freezylib/logging/PolledField.kt
package org.frc1778.freezylib.logging

import org.frc1778.freezylib.util.Measurement

class PolledField<T>(name: String, unit: Measurement.Base, supplier: () -> T) : Field() {

    private val supplier: () -> String
    override lateinit var value: String

    init {
        this.name = name
        this.unit = unit
        setType((supplier.invoke() as Any)::class)
        this.supplier = { supplier().toString() }
    }

    fun pollSupplier() {
        value = supplier()
    }
}
